File: scripts/config-profile.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { ROOT } from "./common.ts";

export const DEFAULT_CONFIG = join(ROOT, "config/default.json");
export const DEFAULT_ENV = join(ROOT, ".env");
export class ProfileError extends Error {}
export type Json =
  null | boolean | number | string | Json[] | { [key: string]: Json };
export interface Profile {
  config: Record<string, Json>;
  payload: string;
}
export function object(value: unknown): value is Record<string, Json> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
function read(path: string, label: string): string {
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(readFileSync(path));
  } catch {
    throw new ProfileError(`Cannot read ${label} (values omitted)`);
  }
}
export function loadEnv(path?: string): Record<string, string> {
  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(
      readFileSync(path ?? DEFAULT_ENV),
    );
  } catch (error) {
    if (
      path === undefined &&
      (error as NodeJS.ErrnoException).code === "ENOENT"
    )
      return {};
    throw new ProfileError("Cannot read environment file");
  }
  const values: Record<string, string> = Object.create(null);
  for (let line of text.split(/\r?\n/)) {
    line = line.trim();
    if (!line || line.startsWith("#")) continue;
    const assignment = /^([A-Za-z_][A-Za-z0-9_]*)\s*=\s*(.*)$/.exec(line);
    if (!assignment || Object.hasOwn(values, assignment[1]))
      throw new ProfileError("Invalid or duplicate environment assignment");
    let value = assignment[2].trim();
    if (value.startsWith('"') || value.startsWith("'")) {
      if (value.length < 2 || value.at(-1) !== value[0])
        throw new ProfileError("Unclosed environment value quote");
      value = value.slice(1, -1);
    }
    values[assignment[1]] = value;
  }
  return values;
}
export function parseJson(text: string): Json {
  try {
    // Inspect container depth and decoded object keys before JSON.parse erases duplicates.
    const containers: (Set<string> | null)[] = [];
    const tokens = text.matchAll(/"(?:[^"\\]|\\[\s\S])*"|[{}\[\]]/g);
    for (const token of tokens) {
      const value = token[0];
      if (value === "{" || value === "[") {
        containers.push(value === "{" ? new Set() : null);
        if (containers.length > 8) throw new Error();
      } else if (value === "}" || value === "]") containers.pop();
      else if (/^\s*:/.test(text.slice(token.index + value.length))) {
        const keys = containers.at(-1);
        const key: string = JSON.parse(value);
        if (!keys || keys.has(key)) throw new Error();
        keys.add(key);
      }
    }
    return JSON.parse(text, (_key, value: Json) => {
      if (typeof value === "number" && !Number.isFinite(value))
        throw new Error();
      return value;
    }) as Json;
  } catch {
    throw new ProfileError(
      "Invalid, duplicate, or excessively nested configuration JSON (values omitted)",
    );
  }
}
export function expand(value: Json, environment: NodeJS.ProcessEnv): Json {
  if (Array.isArray(value))
    return value.map((item) => expand(item, environment));
  if (object(value))
    return Object.fromEntries(
      Object.entries(value).map(([key, item]) => {
        if (key.includes("${"))
          throw new ProfileError(
            "Placeholders are allowed only in JSON string values",
          );
        return [key, expand(item, environment)];
      }),
    );
  if (typeof value !== "string") return value;
  const result = value.replace(
    /\$\{([A-Za-z_][A-Za-z0-9_]*)\}/g,
    (_match, name: string) => {
      if (!Object.hasOwn(environment, name) || environment[name] === undefined)
        throw new ProfileError(
          "Missing referenced environment variable (values omitted)",
        );
      return environment[name];
    },
  );
  if (result.includes("${"))
    throw new ProfileError("Unresolved or unsupported environment placeholder");
  if (!result.isWellFormed())
    throw new ProfileError("Invalid Unicode in configuration");
  return result;
}
export function loadProfile(
  path = DEFAULT_CONFIG,
  envFile?: string,
  environ = process.env,
): Profile {
  const config = expand(parseJson(read(path, "configuration profile")), {
    ...loadEnv(envFile),
    ...environ,
  });
  if (!object(config))
    throw new ProfileError("Configuration must be a JSON object");
  // Firmware owns room and policy validation; these are only cheap framing checks.
  for (const [key, value] of Object.entries(config)) {
    const valid = ["wifi_ssid", "wifi_password", "apple_region"].includes(key)
      ? typeof value === "string"
      : key === "read_only"
        ? typeof value === "boolean"
        : key === "rooms"
          ? object(value)
          : false;
    if (!valid)
      throw new ProfileError(
        "Unknown configuration field or invalid top-level type",
      );
  }
  const payload = JSON.stringify(config);
  if (Buffer.byteLength(payload) > 4088)
    throw new ProfileError("Resolved configuration exceeds 4088 UTF-8 bytes");
  return { config, payload };
}

File: scripts/config_profile.py
"""Load environment profiles and expand JSON string values, entirely on the host."""
import json
import os
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_CONFIG = ROOT / 'config/default.json'
DEFAULT_ENV = ROOT / '.env'
VARIABLE = re.compile(r'\$\{([A-Za-z_][A-Za-z0-9_]*)\}')
KEY = re.compile(r'[A-Za-z_][A-Za-z0-9_]*')


class ProfileError(ValueError):
    """Messages must never contain profile text, environment values, or device output."""


def add_profile_arguments(parser, default=DEFAULT_CONFIG):
    parser.add_argument('--config', type=Path, default=default,
                        help='Environment profile path (independent of board and filename)')
    parser.add_argument('--env-file', type=Path,
                        help='Local KEY=VALUE secrets file (default: repo-root .env)')


def load_env(path=None):
    selected = DEFAULT_ENV if path is None else Path(path)
    try:
        text = selected.read_text(encoding='utf-8')
    except FileNotFoundError:
        if path is None:
            return {}
        raise ProfileError('Environment file not found') from None
    except (OSError, UnicodeError):
        raise ProfileError('Cannot read environment file') from None
    values = {}
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        key, separator, value = line.partition('=')
        key, value = key.strip(), value.strip()
        if not separator or not KEY.fullmatch(key) or key in values:
            raise ProfileError('Invalid or duplicate environment assignment')
        if value.startswith(('"', "'")):
            if len(value) < 2 or value[-1] != value[0]:
                raise ProfileError('Unclosed environment value quote')
            value = value[1:-1]
        values[key] = value
    return values


def unique_object(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            raise ProfileError('Duplicate configuration key')
        result[key] = value
    return result


def reject_constant(_):
    raise ProfileError('Non-finite configuration number')


def parse_json(text):
    try:
        return json.loads(text, object_pairs_hook=unique_object, parse_constant=reject_constant)
    except (ValueError, RecursionError):
        raise ProfileError('Invalid or duplicate configuration JSON (values omitted)') from None


def expand(value, environment, depth=0):
    if isinstance(value, (dict, list)):
        if depth >= 8:
            raise ProfileError('Configuration nesting exceeds eight containers')
        if isinstance(value, dict):
            if any('${' in key for key in value):
                raise ProfileError('Placeholders are allowed only in JSON string values')
            return {key: expand(item, environment, depth + 1) for key, item in value.items()}
        return [expand(item, environment, depth + 1) for item in value]
    if not isinstance(value, str):
        return value

    def substitute(match):
        name = match.group(1)
        if name not in environment:
            raise ProfileError('Missing referenced environment variable (values omitted)')
        return environment[name]

    resolved = VARIABLE.sub(substitute, value)
    if '${' in resolved:
        raise ProfileError('Unresolved or unsupported environment placeholder')
    return resolved


def validate_shape(config):
    # Only host framing/top-level checks. Room/mode semantics belong to firmware.
    if not isinstance(config, dict):
        raise ProfileError('Configuration must be a JSON object')
    strings = {'wifi_ssid', 'wifi_password', 'apple_region'}
    for key, value in config.items():
        if key in strings:
            valid = isinstance(value, str)
        elif key == 'read_only':
            valid = isinstance(value, bool)
        elif key == 'rooms':
            valid = isinstance(value, dict)
        else:
            raise ProfileError('Unknown configuration field')
        if not valid:
            raise ProfileError('Invalid configuration field type')


def load_profile(path=DEFAULT_CONFIG, env_file=None, environ=None):
    try:
        text = Path(path).read_text(encoding='utf-8')
    except (OSError, UnicodeError):
        raise ProfileError('Cannot read configuration profile') from None
    environment = load_env(env_file)
    environment.update(os.environ if environ is None else environ)
    # Parse structure first so quotes/backslashes in credentials cannot inject JSON.
    config = expand(parse_json(text), environment)
    validate_shape(config)
    try:
        payload = json.dumps(config, separators=(',', ':'), ensure_ascii=False, allow_nan=False)
        size = len(payload.encode('utf-8'))
    except (ValueError, UnicodeError):
        raise ProfileError('Invalid resolved configuration JSON (values omitted)') from None
    if size > 4088:
        raise ProfileError('Resolved configuration exceeds 4088 UTF-8 bytes')
    return config, payload
